sound/soc/soc-pcm.c

```diff
@@ -2789,6 +2788,114 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
 	return 0;
 }
 
+/*
+ * create a virtual FE DAI link
+ * Virtual FE DAI links are used in hostless pipelines
+ * to enable the codecs when the pipeline is triggered
+ */
+int soc_dpcm_vfe_new(struct snd_soc_card *card, int index,
+		     const char *link_name, const char *cpu_dai_name,
+		     const char *platform_name)
+{
+	struct snd_soc_dai_link *link;
+
+	link = kzalloc(sizeof(*link), GFP_KERNEL);
+	if (!link)
+		return -ENOMEM;
+
+	dev_dbg(card->dev, "ASoC: adding new virtual FE DAI link %s\n",
+		link_name);
+
+	/* define virtual FE DAI link */
+	link->virtual = 1;
+	link->name = link_name;
+	link->id = 1;
+	link->cpu_dai_name = cpu_dai_name;
+	link->platform_name = platform_name;
+	link->codec_name = "snd-soc-dummy";
+	link->codec_dai_name = "snd-soc-dummy-dai";
+	link->num_codecs = 1;
+
+	/* allocate memory for link codecs */
+	link->codecs = devm_kzalloc(card->dev,
+				    sizeof(struct snd_soc_dai_link_component),
+				    GFP_KERNEL);
+	if (!link->codecs)
+		return -ENOMEM;
+
+	link->codecs[0].name = link->codec_name;
+	link->codecs[0].dai_name = link->codec_dai_name;
+
+	/* enable DPCM */
+	link->dynamic = 1;
+
+	/*TODO: check if we need to handle capture for virtual FE */
+	link->dpcm_playback = 1;
+
+	link->dobj.index = index;
+	link->dobj.type = SND_SOC_DOBJ_DAI_LINK;
+
+	/* add virtual dai link to card dai link list */
+	snd_soc_add_dai_link(card, link);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(soc_dpcm_vfe_new);
+
+/* free virtual FE DAI link */
+int soc_dpcm_vfe_free(struct snd_soc_card *card)
```

Comment:
we need to individually delete these link by link.

Reply:
Is the idea then that the widget that created the virtual fe, will delete it?
In that case, can I associate the link with the comp by adding it to the snd_sof_control associated with the comp?

Reply:
FIxed to delete vfe by name